control-plane/agents/core/src/node/specs.rs

```diff
@@ -140,4 +141,29 @@ impl ResourceSpecsLocked {
             })
             .collect()
     }
+
+    /// Drain the node with the given ID.
+    /// Return the NodeSpec after draining.
+    pub(crate) async fn drain_node(
+        &self,
+        registry: &Registry,
+        node_id: &NodeId,
+        label: String,
+    ) -> Result<NodeSpec, SvcError> {
+        let node = self.get_locked_node(node_id)?;
+        let drained_node_spec = {
+            let mut locked_node = node.lock();
+            // Do not allow the same label to be applied more than once.
+            if locked_node.has_cordon_label(label.clone()) {
+                return Err(SvcError::CordonLabel {
+                    node_id: node_id.to_string(),
+                    label,
+                });
+            }
+            locked_node.drain(label);
+            locked_node.clone()
+        };
+        registry.store_obj(&drained_node_spec).await?;
```

Comment:
now that we're modifying the node spec for cordon and drain we should probably impl the SpecOperations trait for the Node, and update the spec through that interface.
This should be a separate task though.